### ISSUE TITLE
ArrayIndexOutOfBoundsException when serializing strings with ASCII escapable characters

### ISSUE THREAD

**Issue by raghav-deepsource** [Kotlin/kotlinx.serialization#1530]
the JsonStringBuilder class currently throws an `ArrayIndexOutOfBoundsException` for inputs such as: 
```
"\naaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
 ```
omit the quotes and render the \n as a newline, then feed this to JsonStringBuilder to trigger this bug.

**To Reproduce**
```kotlin


fun toHexChar(i: Int) : Char {
    val d = i and 0xf
    return if (d < 10) (d + '0'.toInt()).toChar()
    else (d - 10 + 'a'.toInt()).toChar()
}

var array = CharArray(128)
var size = 0


fun ensureAdditionalCapacity(expected: Int) {
    ensureTotalCapacity(size + expected)
}

fun ensureTotalCapacity(newSize: Int) {
    if (array.size <= newSize) {
        array = array.copyOf(newSize.coerceAtLeast(size * 2))
    }
}

val ESCAPE_MARKERS: ByteArray = ByteArray(93).apply {
    for (c in 0..0x1f) {
        this[c] = 1.toByte()
    }
    this['"'.code] = '"'.toByte()
    this['\\'.code] = '\\'.toByte()
    this['\t'.code] = 't'.toByte()
    this['\b'.code] = 'b'.toByte()
    this['\n'.code] = 'n'.toByte()
    this['\r'.code] = 'r'.toByte()
    this[0x0c] = 'f'.toByte()
}

val ESCAPE_STRINGS: Array<String?> = arrayOfNulls<String>(93).apply {
    for (c in 0..0x1f) {
        val c1 = toHexChar(c shr 12)
        val c2 = toHexChar(c shr 8)
        val c3 = toHexChar(c shr 4)
        val c4 = toHexChar(c)
        this[c] = "\\u$c1$c2$c3$c4"
    }
    this['"'.code] = "\\\""
    this['\\'.code] = "\\\\"
    this['\t'.code] = "\\t"
    this['\b'.code] = "\\b"
    this['\n'.code] = "\\n"
    this['\r'.code] = "\\r"
    this[0x0c] = "\\f"
}

fun appendQuoted(string: String) {
    ensureAdditionalCapacity(string.length + 2)
    val arr = array
    var sz = size
    arr[sz++] = '"'
    val length = string.length
    string.toCharArray(arr, sz, 0, length)
    for (i in sz until sz + length) {
        val ch = arr[i].toInt()
        println("${arr[i]} $ch")
        // Do we have unescaped symbols?
        if (ch < ESCAPE_MARKERS.size && ESCAPE_MARKERS[ch] != 0.toByte()) {
            // Go to slow path
                println("do the thing at pos $i of total ${arr.size}")
            return appendStringSlowPath(i - sz, i, string)
        }
    }
    // Update the state
    sz += length
    arr[sz++] = '"'
    size = sz
}

fun appendStringSlowPath(firstEscapedChar: Int, currentSize: Int, string: String) {
    var sz = currentSize
    for (i in firstEscapedChar until string.length) {
        val ch = string[i].toInt()
        println("pos $i newpos $sz character ${string[i]} code $ch")
        // Do we have unescaped symbols?
        if (ch < ESCAPE_MARKERS.size) {
            /*
            * Escape markers are populated for backslash-escaped symbols.
            * E.g. ESCAPE_MARKERS['\b'] == 'b'.toByte()
            * Everything else is populated with either zeros (no escapes)
            * or ones (unicode escape)
            */
            when (val marker = ESCAPE_MARKERS[ch]) {
                0.toByte() -> {
                    array[sz++] = ch.toChar()
                }
                1.toByte() -> {
                    val escapedString = ESCAPE_STRINGS[ch]!!
                    ensureTotalCapacity(sz + escapedString.length)
                    escapedString.toCharArray(array, sz, 0, escapedString.length)
                    sz += escapedString.length
                }
                else -> {

                    println("skip twice at pos $i newpos $sz character ${ch.toChar()} code $ch")
                    array[sz] = '\\'
                    array[sz + 1] = marker.toChar()
                    sz += 2
                }
            }
        } else {
            array[sz++] = ch.toChar()
        }
    }
    println("final pos $sz")
    array[sz++] = '"'
    size = sz
}

val str = """
""" + "a".repeat(125)

appendQuoted(str)

println(String(array, 0, size))

```

this is an mvp script with code from JsonStringBuilder which demonstrates this bug as well. note that I have added some print lines to log state but have not changed the code much other than that.

cause seems to be that while the array is extended as necessary when dealing with unicode, this same behavior does not occur when escapable ascii characters are present.

**Expected behavior**

the string should have been serialized successfully.

**Environment**
 - Kotlin version: 1.5.0
 - Library version: 1.2.0
 - Kotlin platforms: JVM
 - Gradle version: N/A
 - IDE version (if bug is related to the IDE) N/A

**Comment by shanshin** [comment]
Is this issue reproduced in library version 1.2.1?

**Comment by raghav-deepsource** [comment]
not sure.

**Comment by raghav-deepsource** [comment]
I believe this bug will occur on master

**Comment by shanshin** [comment]
The issue is similar to the #1504, #1464 - it should be fixed in 1.2.1

**Comment by raghav-deepsource** [comment]
yup, seems so.
